lib/tasks_private/instance_types.rake

```diff
@@ -1,215 +1,277 @@
+      end
+    end
+
+    discontinued_types, types_list = isolated do
+      previous_list = []
+      versions = get_gh_data("contents/#{SDK_DATA_DIR}")
+      versions = versions.lazy.select { |v| v['type'] == 'dir' }.map { |v| v['name'] }.sort
+      all_types, latest = versions.reduce([[], []]) do |(all, _latest), version|
+        data = get_gh_file("#{SDK_DATA_DIR}/#{version}/#{SDK_FILE_NAME}")
+        data = data['shapes']['InstanceType']['enum']
+        old_minus_new = previous_list - data
+        new_minus_old = data - previous_list
+        log_data("AWS EC2 SDK data version #{version} removed types:", old_minus_new)
+        log_data("AWS EC2 SDK data version #{version} added types:", new_minus_old, :level => :info)
+        previous_list = data
+        [all | data, data]
+      end
+      [all_types - latest, latest]
+    end
+
+    ## get, parse, and sort data
+
+    types_data, collecting_warnings, parsing_warnings = isolated do
+      products_data, collecting_warnings = AwsProductsDataCollector.new(
+        :service_name       => 'AmazonEC2',
+        :product_families   => 'Compute Instance', # 'Dedicated Host' == bare metal: "m5", "p3", etc.
+        :product_attributes => AwsInstanceDataParser::REQUIRED_ATTRIBUTES,
+        :folding_attributes => 'instanceType',
+        :mutable_attributes => %w(currentVersion currentGeneration),
+      ).result
+
+      parsing_warnings = {}
+
+      types_data = products_data.map do |product_data|
+        instance_data, warnings = AwsInstanceDataParser.new(product_data).result
+        parsing_warnings.merge!(warnings) { |_, old, new| old + new }
+        [product_data['instanceType'], instance_data.deep_dup]
+      end.to_h
 
-  puts <<EOT
-Created ./instance_types.rb with #{new_count} new, #{deprecated_count} newly deprecated and #{discontinued_count} new discontinued instances.
-Now copy it to the existing one:
+      [types_data, collecting_warnings, parsing_warnings]
+    end
+
+    ## consider previous data
+
+    old_types_data = YAML.load_file(out_file)
+    types_list = types_list | old_types_data.keys | types_data.keys
+
+    default_type = old_types_data.find { |_, data| data[:default] }.first
+    discontinued_types |= (old_types_data.keys - types_data.keys)
+    types_data.merge!(old_types_data.slice(*discontinued_types))
+
+    ## postprocess
+
+    types_data.sort_by! do |instance_type, instance_data|
+      instance_data.except!(:default, :deprecated, :discontinued, :disabled)
+      instance_data[:default] = true if instance_type == default_type
+      if discontinued_types.include?(instance_type) || !instance_data[:current_version]
+        instance_data[:discontinued] = true
+        instance_data[:disabled] = true
+      end
+      if instance_data[:current_version] && !instance_data[:current_generation]
+        instance_data[:deprecated] = true
+      end
+      types_list.index(instance_type) || 1_000_000
```

Comment:
How do unindexed types get sorted amongst themselves?  I would expect by instance_type name or something.

Reply:
They should keep previous order. And it is impossible to sort by numbers together with the strings.

Reply:
It's possible, you just need a multi-level sort...

```ruby
[types_list.index(instance_type) || 1_000_000, instance_data[:instance_type]
```

If you return that from the sort_by! block it will do multi-level sort on previous index, then name.

Reply:
That being said, it's much more intuitive to simplify this to just sort by name instead of previous index.  New values will get inserted where they belong and the diff will still only include the new stuff.

Reply:
1) It complicates code and readability, and (maybe less important, but...) it will decrease speed and increase memory usage.
2) Types list order can't be sorted alphabetically, that's why I'm taking types list from AWS GitHub repo, to make it consistent with AWS lists.

Reply:
> It complicates code and readability, and (maybe less important, but...) it will decrease speed and increase memory usage.

That should not be a concern, unless you are saying it will take *hours* and use *gigs* of memory.  This is a one-off execution.

> Types list order can't be sorted alphabetically (it leads to bad end user experience), that's why I'm taking types list from AWS GitHub repo, to make it consistent with AWS lists.

I don't understand the bad end user experience.  The sorting in the file is strictly for diff purposes and the end user will never see that sorting.  Unless you mean the dev running this rake task as the end-user?